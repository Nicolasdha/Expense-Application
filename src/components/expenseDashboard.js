import React from 'react';


import ExpenseList from './expenseList';
import ExpenseListFilters from './expenseListFilters';
import ExpensesSummary from './expensesSummary';


const ExpenseDashboardPage = () => {
   
return ( 
    <div>
        {<ExpenseListFilters/>}
        {<ExpensesSummary/>}
        {<ExpenseList/>}
    </div>
    );
};


export default ExpenseDashboardPage;
