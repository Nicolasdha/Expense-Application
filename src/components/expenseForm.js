import React from 'react';
import moment from'moment';
import { SingleDatePicker } from 'react-dates';
import 'react-dates/initialize';



class ExpenseForm extends React.Component {

    constructor(props){ 
      super(props);
        this.state = {
            description: props.match ? props.match.description : '',
            note: props.match ? props.match.note : '',
            amount: props.match ? (props.match.amount / 100).toString() : '',
            createdAt: props.match ? moment(props.match.createdAt) : moment(),
            calandarFocused: false,
            error: '',
        } 
    }

    
    onDescriptionChange = (e) => {
        const description = e.target.value
        this.setState(() =>({ description }))
    };

    onNoteChange = (e) =>{
        const note = e.target.value
        this.setState(() =>({ note }))
    };

    onAmountChange = (e) =>{
        const amount = e.target.value;
        if(!amount || amount.match(/^\d{1,}(\.\d{0,2})?$/g)){
            this.setState(()=>({ amount }))
        }
    };

    onDateChange = (createdAt) =>{
        if(createdAt){
            this.setState(()=> ({ createdAt }))
        }
    };

    onFocusChange = ( { focused } ) => {
        this.setState(()=>({ calandarFocused: focused}))
    };

    onSubmit = (e) =>{
        e.preventDefault();
        if(!this.state.description || !this.state.amount){
            this.setState(()=> ({error: 'Please provide Description and Amount'}));
        } else{
            this.setState(()=> ({error: ''}));
            this.props.onSubmit({
                description: this.state.description,
                amount: parseFloat(this.state.amount, 10) * 100,
                note: this.state.note,
                createdAt: this.state.createdAt.valueOf()
            })
        }
    }

    render(){
        return (
            <form className="form" onSubmit={this.onSubmit}>
                {this.state.error && <p className="form__error">{this.state.error}</p>}

                <input
                    className="text-input"
                    type="text" 
                    placeholder="Description" 
                    value = {this.state.description} 
                    onChange = {this.onDescriptionChange}
                    autoFocus
                />
                <input 
                    type="text" 
                    className="text-input"
                    placeholder="Amount"
                    value = {this.state.amount}
                    onChange = {this.onAmountChange}
                />
                <SingleDatePicker 
                    date = {this.state.createdAt}
                    onDateChange = {this.onDateChange}
                    focused = {this.state.calandarFocused}
                    onFocusChange = {this.onFocusChange}
                    numberOfMonths = {1}
                    isOutsideRange = {(day)=>false}
                />
                <textarea 
                    className="textarea"
                    placeholder="Add a note (optional)"
                    value={this.state.note}
                    onChange = {this.onNoteChange}
                ></textarea>
                <div>
                    <button className="button">Save Expense</button>
                </div>
            </form>
        );
    };
};

export default ExpenseForm;
