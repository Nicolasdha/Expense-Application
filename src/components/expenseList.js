import React from 'react';
import { connect } from 'react-redux';
import ExpenseListItem from './expenseListItem';
import getVisibleExpenses from '../selectors/expenses';



export const ExpenseList = (props) => {
    
    return (
        <div>
            {
                props.expenses.length === 0 ? (
                    <p>Please add expense!</p>) : (
                        props.expenses.map((expense) => <ExpenseListItem key={expense.id} {...expense}/>) 
                    )
            }
        </div>
)};

const mapStoreToProps = (state) =>{
    return {
        expenses: getVisibleExpenses(state.expenses, state.filters)
    };
};

export default connect(mapStoreToProps)(ExpenseList);
