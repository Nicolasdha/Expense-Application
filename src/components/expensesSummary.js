import React from 'react';
import numeral from 'numeral';
import { Link } from 'react-router-dom'
import { connect } from 'react-redux';
import getVisibleExpenses from '../selectors/expenses';

export const ExpensesSummary = (props) => {
    let expensesForTotal = [];
    let reducer = (accumulator, currentValue) => accumulator + currentValue;


    props.visibleExpenses.map((expense) =>{
        expensesForTotal.push(expense.amount / 100)
    });

    const total = expensesForTotal.length === 0 ? null : expensesForTotal.reduce(reducer);
    const expenseWord = props.expenseCount === 1 ? 'expense' : 'expenses';
    const formattedExpensesTotal = numeral(total).format('$0,0.00');



    return (
    <div className = "page-header">
        <div className="content-container">
            <h1 className="page-header__title">Viewing <span>{props.expenseCount} </span>{expenseWord} totaling <span>{formattedExpensesTotal}</span></h1>
            <hr></hr>
            <div className = "page-header__actions">
                <Link className="button" to='/create'>Add Expense</Link>
            </div>
        </div>
    </div>
    )
};


const mapStoreToProps = (state) =>{
    const visibleExpenses = getVisibleExpenses(state.expenses, state.filters)
    return {
        visibleExpenses: visibleExpenses,
        expenseCount: visibleExpenses.length,
    }
};


export default connect(mapStoreToProps)(ExpensesSummary);

