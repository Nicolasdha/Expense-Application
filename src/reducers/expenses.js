
// Expenses Reducer

export default (state = [], action ) =>{
    switch(action.type){
        case 'ADD_EXPENSE':
            return [
                ...state,
                action.expense
            ];
        case 'REMOVE_EXPENSE':
            return state.filter(({ id }) => id !== action.id);
        case 'EDIT_EXPENSE':
            return state.map((each) =>{
                if(each.id === action.id){
                    return {
                        ...each,
                        ...action.updates
                    }
                } else {
                    return each
                }
            })
        default:
            return state;
    };
};

