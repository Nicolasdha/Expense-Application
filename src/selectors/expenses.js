import moment from 'moment';
// Get visible state

export default (expenses, {  text, sortBy, startDate, endDate }) =>{
    return expenses.filter((each) =>{
        const createdAtMoment = moment(each.createdAt)
        const startDateMatch = startDate ? startDate.isSameOrBefore(createdAtMoment, 'day') : true;
        const endDateMatch = endDate ? endDate.isSameOrAfter(createdAtMoment, 'day') : true;
        const textMatch = each.description.toLowerCase().includes(text.toLowerCase());

        return startDateMatch && endDateMatch && textMatch;
    }).sort((a,b) =>{
        if(sortBy === 'date'){
            return a.createdAt < b.createdAt ? 1 : -1
        } else if(sortBy ==='amount'){
            return a.amount < b.amount ? 1 : -1
        }
    })
};



