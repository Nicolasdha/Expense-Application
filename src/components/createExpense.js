import React from 'react';
import { connect } from 'react-redux';
import { startAddExpense } from '../actions/expenses';
import ExpenseForm from './expenseForm'



export class CreateExpensePage extends React.Component {

    onSubmit = (expense) => { 
        this.props.startAddExpense(expense);
        this.props.history.push('/');
    };

    render(){
        return (
            <div>
                <h1>Add Expense</h1>
                <ExpenseForm onSubmit = {this.onSubmit}/>
            </div>
        );
    };
};

const mapDispatchToProps = (dispatch) => ({
    startAddExpense: (expense) => dispatch(startAddExpense(expense))
})


export default connect(undefined, mapDispatchToProps)(CreateExpensePage)


// const CreateExpensePage = (props) => (
//     <div>
//         <h1>Add Expense</h1>
//         <ExpenseForm
//             onSubmit = {(expense) => { 
//                 props.onSubmit(expense);
//                 props.history.push('/');
//             }}
//         />
//     </div>
// );




// const CreateExpensePage = (props) => (
//     <div>
//         <h1>Add Expense</h1>
//         <ExpenseForm
//             onSubmit = {(expense) => { 
//                 props.dispatch(addExpense(expense))
//                 props.history.push('/')
//             }}
//         />
//     </div>
// );

// export default connect()(CreateExpensePage)