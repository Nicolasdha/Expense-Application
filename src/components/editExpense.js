import React from 'react';
import { connect } from 'react-redux';
import Modal from 'react-modal'
import { startEditExpense, startRemoveExpense } from '../actions/expenses';
import ExpenseForm from './expenseForm'




export class EditExpensePage extends React.Component {
    state = {
        modalState: undefined
      };

    onSubmit = (expense) =>{ 
            this.props.startEditExpense(this.props.match.id, expense)
            this.props.history.push('/')
    };

    openModal = () => {
        this.setState(()=>({ modalState: true }
            ));
     };
     
     closeModal = () => {
         this.setState(()=>({ modalState: undefined }))
     }

     onRemove = () =>{
        this.setState(()=>({ modalState: undefined }
        ));
        this.props.startRemoveExpense( {id: this.props.match.id} )
        this.props.history.push('/')
      }

    render() {

        return (
            <div>
            <div className="page-header">
                <div className="content-container">
                    <h1 className="page-header__title">Edit Expense</h1>
                </div>
            </div>
                <div className="content-container">
                    <ExpenseForm
                        match = {this.props.match}
                        onSubmit = {this.onSubmit}
                        />
                    <button className="button button--secondary" onClick={this.openModal}>Remove Expense</button>
                    <Modal
                        isOpen={!!this.state.modalState} 
                        contentLabel="Are you sure you want to remove this expense?"
                        appElement={document.getElementById('app')}
                        onRequestClose={this.closeModal}
                        closeTimeoutMS={200} 
                        className="modal"
                    >
                       <div className="modal__titleContainer">
                            <h3 className="modal__title">Remove expense?</h3>
                       </div>
                        <div className="modal__buttonContainer">
                            <button className="button modal--button" onClick={this.onRemove}>Yes</button>
                            <button className="button modal--button" onClick={this.closeModal}>No</button>
                        </div>

                    </Modal>
                </div>
            </div>
        );
    };
};


const mapStoreToProps = (state, props) =>({
    match: state.expenses.find((each)=> each.id === props.match.params.id),
});



const mapDispatchToProps = (dispatch) => ({
    startEditExpense: (id, expense) => dispatch(startEditExpense(id, expense)),
    startRemoveExpense: (id)=> dispatch(startRemoveExpense(id))
});



export default connect(mapStoreToProps, mapDispatchToProps)(EditExpensePage);




// const EditExpensePage = (props) => {
//     const match = props.expenses.find((each)=> each.id === props.match.params.id)
//     return (
//     <div>
//         <ExpenseForm
//             match = {match}
//             onSubmit = {(expense) =>{ 
//                 props.dispatch(startEditExpense(match.id, expense))
//                 props.history.push('/')
//             }}
//             />
//         <button
//             onClick={() => {
//                 const removeConfirmation = window.prompt('Are you sure you want to remove? Type Y to confirm')
//                 if(removeConfirmation === 'yes'.toLowerCase() || 'y'.toLowerCase()){
//                     props.dispatch(startRemoveExpense( {id: match.id} ))
//                     props.history.push('/')
//                 }
//              }}
//         >Remove</button>
//     </div>
//     );
// };
