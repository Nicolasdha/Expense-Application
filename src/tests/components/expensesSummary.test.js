import React from 'react';
import { shallow } from 'enzyme';

import { ExpensesSummary } from '../../components/expensesSummary';
import expenses from '../fixtures/expenses'



test('should render the correct total', () =>{
    const wrapper = shallow(<ExpensesSummary expenseCount={expenses.length} visibleExpenses={expenses}/>)
    expect(wrapper).toMatchSnapshot();
})