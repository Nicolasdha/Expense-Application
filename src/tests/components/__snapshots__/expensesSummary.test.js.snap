// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render the correct total 1`] = `
<div
  className="page-header"
>
  <div
    className="content-container"
  >
    <h1
      className="page-header__title"
    >
      Viewing 
      <span>
        3
         
      </span>
      expenses
       totaling 
      <span>
        $1,996.95
      </span>
    </h1>
    <hr />
    <div
      className="page-header__actions"
    >
      <Link
        className="button"
        to="/create"
      >
        Add Expense
      </Link>
    </div>
  </div>
</div>
`;
